### PR TITLE
6563 keep orders in sync

### LINKED ISSUE #6563: Snapshots: failure seen on stagnet1 pipeline
# Problem encountered
Snapshot failures are being seen on the stagnet1 snapshot-pipeline

We need to investigate and fix.

# Observed behaviour
Snapshots created after block ~40000 fall out of consensus with the following error:
```
[2022-10-21T10:34:00.773Z] 2022-10-21T10:34:00.603Z	ERROR	tendermint	v0/reactor.go:379	Error in validation	{"err": "wrong Block.Header.LastResultsHash.  Expected 2929F9ECA252194F20D5846E6DD23A4B69EEA006433A0A8993F0D0740C3C1D76, got 18E072A5DCBF652CED0043A929675258C5B243533EC043302B5001C48ED0B4BA"}
```

# Expected behaviour
All snapshots should work.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #6563

This PR changes the liquidity engineso that it asks the order-book for orders instead of keeping its own copy of orders-per-liquidity-provider. This will remove a class of bugs that could cause snapshot-restores to fail if the orders did become out-of-sync.

The change has caused some numbers in the integration tests to change slightly, with the difference being traced down to instances were the order-book and the orders in the liquidity engine did become out of sync.

Full system test run shows failures that match with existing overnight run failures: https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests/detail/system-tests/12435/tests

### CODE REVIEW

File: core/execution/liquidity_provision.go [hunk 1]
```diff
@@ -707,6 +709,14 @@ func (m *Market) cancelLiquidityProvision(
 		return err
 	}
 
+	// FIXME(JEREMY): if sorting them is the actual solution
+	// review the implementation to write some eventually more efficient
+	// way to sort this here and make sure that all orders are always
+	// cancelled in the same order
```
Comment (jeremyletang) [inline core/execution/liquidity_provision.go:715]: that probably can go now haha

File: core/liquidity/engine.go [hunk 2]
```diff
@@ -306,6 +261,18 @@ func (e *Engine) ProvisionsPerParty() ProvisionsPerParty {
 	return e.provisions.ProvisionsPerParty
 }
 
+// SaveLPOrders sets LP orders that have been cancelled from the book but we need to know
+// what they were to recalculate and redeploy.
+func (e *Engine) SaveLPOrders() {
+	for _, p := range e.provisions.ProvisionsPerParty {
+		e.lpPartyOrders[p.Party] = e.orderBook.GetOrdersPerParty(p.Party)
```
Comment (jeremyletang) [inline core/liquidity/engine.go:268]: you will have all the order of the LP here, not only those created by the liquidity engine, is that OK?

Comment (jeremyletang) [reply core/liquidity/engine.go:268]: Would you not want to use GetLiquidityOrders(party) here?

File: core/matching/snapshot.go [hunk 3]
```diff
@@ -130,13 +122,10 @@ func (b *OrderBook) LoadState(_ context.Context, payload *types.Payload) ([]type
...
 
-	if orders, ok := b.ordersPerParty[o.Party]; !ok {
-		b.ordersPerParty[o.Party] = map[string]struct{}{
-			o.ID: {},
-		}
-	} else {
-		orders[o.ID] = struct{}{}
+	// reconstruct pegged orders state
+	if o.PeggedOrder != nil {
+		b.peggedOrders[o.ID] = struct{}{}
```
Comment (jeremyletang) [inline core/matching/snapshot.go:129]: why not add this in the add(order) function?

**Comment by jeremyletang** [comment]
> I've done some perf testing on this branch and the number of events it produces and the peak order rate for the number of LP orders is exactly the same as before, no better and no worse. Here is a quick update of the table compared to the previous branch with the LP optimisations in: image

thanks @peterbarrow

File: core/liquidity/engine.go (hunk 2 above)
Comment (wwestgarth) [reply core/liquidity/engine.go]: Yea good spot, changed it.

File: core/matching/snapshot.go (hunk 3 above)
Comment (wwestgarth) [reply core/matching/snapshot.go]: Moved it into `add()`.

**Review by davidsiska-vega — APPROVED** [review]
I've looked at the `*.feature` files and numbers within and I don't see anything too troublesome.

**Review by ze97286 — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**